clustermesh-apiserver/mcsapi-coredns-cfg/root.go

```diff
@@ -0,0 +1,175 @@
+	applycorev1 "k8s.io/client-go/applyconfigurations/core/v1"
+
+	"github.com/cilium/cilium/pkg/annotation"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/version"
+)
+
+func NewCmd(h *hive.Hive) *cobra.Command {
+	rootCmd := &cobra.Command{
+		Use:   "mcsapi-coredns-cfg",
+		Short: "Automatically configure CoreDNS with recommended MCS-API settings",
+		Run: func(cmd *cobra.Command, args []string) {
+			// slogloggercheck: it has been initialized in the PreRun function.
+			if err := h.Run(logging.DefaultSlogLogger); err != nil {
+				// slogloggercheck: log fatal errors using the default logger before it's initialized.
+				logging.Fatal(logging.DefaultSlogLogger, err.Error())
+			}
+		},
+		PreRun: func(cmd *cobra.Command, args []string) {
+			option.Config.SetupLogging(h.Viper(), "mcsapi-coredns-cfg")
+
+			// slogloggercheck: the logger has been initialized in the SetupLogging call above
+			log := logging.DefaultSlogLogger.With(logfields.LogSubsys, "mcsapi-coredns-cfg")
+
+			option.Config.Populate(log, h.Viper())
+			option.LogRegisteredSlogOptions(h.Viper(), log)
+			log.Info("Cilium MCS-API CoreDNS auto configuration", logfields.Version, version.Version)
+		},
+	}
+
+	h.RegisterFlags(rootCmd.Flags())
+	rootCmd.AddCommand(h.Command())
+
+	return rootCmd
+}
+
+func configureCoreDNS(jg job.Group, logger *slog.Logger, shutdowner hive.Shutdowner, client client.Clientset, config coreDNSConfig) {
+	jg.Add(job.OneShot("mcsapi-coredns-cfg", func(ctx context.Context, _ cell.Health) (err error) {
+		logger.Info("Starting MCS-API CoreDNS auto configuration")
+		defer func() {
+			shutdowner.Shutdown(hive.ShutdownWithError(err))
+		}()
+
+		if !client.IsEnabled() {
+			err = fmt.Errorf("Kubernetes client is not enabled, cannot configure CoreDNS")
+			logger.Error(err.Error())
+			return
+		}
+
+		var coreDNSConfigmap *corev1.ConfigMap
+		coreDNSConfigmap, err = client.CoreV1().ConfigMaps(config.CoreDNSNamespace).Get(ctx, config.CoreDNSConfigMapName, metav1.GetOptions{})
+		if err != nil {
+			logger.Error("Failed to get CoreDNS ConfigMap", logfields.Error, err)
+			return
+		}
+
+		var corefile string
+		var found bool
+		if corefile, found = coreDNSConfigmap.Data["Corefile"]; !found {
+			err = fmt.Errorf("Corefile not found in ConfigMap %s/%s", config.CoreDNSNamespace, config.CoreDNSConfigMapName)
+			logger.Error(err.Error())
+			return
+		}
+
+		if corefile, err = updateCorefile(config.CoreDNSClusterDomain, config.CoreDNSClustersetDomain, corefile); err != nil {
+			logger.Error("Failed to update CoreDNS Corefile", logfields.Error, err)
+			return
+		}
+		if corefile == "" {
+			logger.Info("CoreDNS might already have MCS-API configuration, skipping configuration")
+			return
+		}
+		coreDNSConfigmap.Data["Corefile"] = corefile
```

Comment:
I wonder if it could make sense to save the previous version under another key (e.g., `Corefile.bak`), just to make it easy to restore it in case of necessity.